### PR TITLE
1.15 Item Drop Event

### PR DESCRIPTION
**Pull request by Draylar**
1.15 extension of #260 due to butchered git history. See original PR for discussion & details.

### CODE REVIEW

File: fabric-events-interaction-v0/src/main/java/net/fabricmc/fabric/api/event/player/PlayerDropItemCallback.java [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+
+import net.fabricmc.fabric.api.event.Event;
+import net.fabricmc.fabric.api.event.EventFactory;
+
+/**
+ * Callback for the player dropping an item entity.
+ * Is hooked in before the item is dropped.
+ *
+ * <p>Upon return:
+ * <ul><li>SUCCESS cancels further processing and drops the item.
```
Comment (liach) [inline fabric-events-interaction-v0/src/main/java/net/fabricmc/fabric/api/event/player/PlayerDropItemCallback.java:31]: Note there is also a `CONSUME` value.

**Comment by liach** [comment]
I disagree with this event. The event imo is better at where this method is called (for different use cases) instead of patching this method and breaking unintended users. And this method is not broad enough either to cover purposes like replacing dropped item's type, etc.

**Comment by Draylar** [comment]
> I disagree with this event. The event imo is better at where this method is called (for different use cases) instead of patching this method and breaking unintended users. And this method is not broad enough either to cover purposes like replacing dropped item's type, etc.

By "unintended users," are you referring to things like death inventory drops?

The standard Q drop stems like this:
  - `MinecraftClient#dropSelectedItem`, ln1433
  - `PlayerEntity#dropSelectedItem`
  - `PlayerEntity#dropItem(ItemStack, boolean, boolean)`
     - handled by `ServerPlayerEntity#dropItem`, which is the current target

This is probably the primary target here; others would be pressing Q in an inventory, and dropping an item to the side of a container screen. Not sure how we could split that up effectively unless there were multiple mixins in each primary area that then called a single callback.

For changing the dropped item itself (which I fully agree with), I could change the event to return a `TypedActionResult`. Would this be the best way to go about it, and do you think it's appropriate for a general drop callback?

**Comment by Draylar** [comment]
I was just looking through the different options we have for this again. Assuming the primary target is items dropped through Q (in-inventory or while held), `MinecraftClient#handleInputEvents#keyDrop.wasPressed()` is a good target. It only covers those two targets, while not butting into things like death item drops. If we also wanted to handle dropping items by clicking on the outside of containers, we could target several `dropItem` calls in `Container`. The problem, again, is that it would be easy to miss edge cases (I'm sure there are mechanics that would allow for dropping the item and not triggering the callback).

I'll re-setup the event to allow for modifying the dropped item anyways; just want to discuss these specifics first.

**Pull request closed without merging**